lib/internal/fetch/headers.js

```diff
@@ -0,0 +1,291 @@
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+
+    if (this[kHeadersList][index] === normalizedName) {
+      ArrayPrototypeSplice(this[kHeadersList], index, 2);
+    }
+  }
+
+  get(name) {
+    if (!isHeaders(this)) {
+      throw new ERR_INVALID_THIS('Headers');
+    }
+
+    const normalizedName = normalizeAndValidateHeaderName(name);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+
+    if (this[kHeadersList][index] === normalizedName) {
+      return this[kHeadersList][index + 1];
+    }
+
+    return null;
+  }
+
+  has(name) {
+    if (!isHeaders(this)) {
+      throw new ERR_INVALID_THIS('Headers');
+    }
+
+    const normalizedName = normalizeAndValidateHeaderName(name);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+
+    return this[kHeadersList][index] === normalizedName;
+  }
+
+  set(name, value) {
+    if (!isHeaders(this)) {
+      throw new ERR_INVALID_THIS('Headers');
+    }
+
+    const normalizedName = normalizeAndValidateHeaderName(name);
+    const normalizedValue = normalizeAndValidateHeaderValue(name, value);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+    if (this[kHeadersList][index] === normalizedName) {
+      this[kHeadersList][index + 1] = normalizedValue;
+    } else {
+      ArrayPrototypeSplice(
+        this[kHeadersList], index, 0, normalizedName, normalizedValue);
+    }
+  }
+
+  * keys() {
+    if (!isHeaders(this)) {
+      throw new ERR_INVALID_THIS('Headers');
+    }
+
+    for (let index = 0; index < this[kHeadersList].length; index += 2) {
+      yield this[kHeadersList][index];
+    }
+  }
+
+  * values() {
+    if (!isHeaders(this)) {
+      throw new ERR_INVALID_THIS('Headers');
+    }
+
+    for (let index = 1; index < this[kHeadersList].length; index += 2) {
+      yield this[kHeadersList][index];
+    }
+  }
+
+  * entries() {
+    if (!isHeaders(this)) {
+      throw new ERR_INVALID_THIS('Headers');
+    }
+
+    for (let index = 0; index < this[kHeadersList].length; index += 2) {
+      yield [this[kHeadersList][index], this[kHeadersList][index + 1]];
+    }
+  }
+
+  forEach(callback, thisArg) {
+    if (!isHeaders(this)) {
+      throw new ERR_INVALID_THIS('Headers');
+    }
+
+    for (let index = 0; index < this[kHeadersList].length; index += 2) {
+      callback.call(
+        thisArg,
+        this[kHeadersList][index + 1],
+        this[kHeadersList][index],
+        this
+      );
+    }
+  }
+
+  [inspect.custom] () {
+    return this[kHeadersList]
+  }
+}
+
+Headers.prototype[SymbolIterator] = Headers.prototype.entries;
+
+ObjectDefineProperties(Headers.prototype, {
+  append: { enumerable: true },
+  delete: { enumerable: true },
+  get: { enumerable: true },
+  has: { enumerable: true },
+  set: { enumerable: true },
+  keys: { enumerable: true },
+  values: { enumerable: true },
+  entries: { enumerable: true },
+  forEach: { enumerable: true },
```

Comment:
```js
const {
  kEnumerableProperty,
} = require('internal/webstreams/util');
```
```suggestion
  append: kEnumerableProperty,
  delete: kEnumerableProperty,
  get: kEnumerableProperty,
  has: kEnumerableProperty,
  set: kEnumerableProperty,
  keys: kEnumerableProperty,
  values: kEnumerableProperty,
  entries: kEnumerableProperty,
  forEach: kEnumerableProperty,
```
It doesn't have to happen in this PR, but it may be worth moving this utility somewhere that is not directly related to webstreams – although `fetch` and Web Streams are not too far away.